test/integration/controller_binding_test.go

```diff
@@ -0,0 +1,407 @@
+	fakeosb "github.com/pmorie/go-open-service-broker-client/v2/fake"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	"github.com/kubernetes-incubator/service-catalog/test/util"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// TestCreateServiceInstanceNonExistentClusterServiceClassOrPlan tests that a ServiceInstance gets
+// a Failed condition when the service class or service plan it references does not exist.
+func TestCreateServiceBinding(t *testing.T) {
+	cases := []struct {
+		name                string
+		instanceName        *string
+		instanceClassName   *string
+		instancePlanName    *string
+		instanceNotReady    bool
+		asyncForInstances   bool
+		asyncForBindings    bool
+		expectedErrorReason string
+		expectedFailure     bool
+		nonbindablePlan     bool
+		duplicateParameters bool
+	}{
+		{
+			name: "happy",
+		},
+		{
+			name:                "non-existent service instance name",
+			instanceName:        strPtr("nothereinstance"),
+			expectedErrorReason: "ReferencesNonexistentInstance",
+		},
+		{
+			name:                "invalid service instance name",
+			instanceName:        strPtr(""),
+			expectedErrorReason: "ReferencesNonexistentInstance",
+			expectedFailure:     true,
+		},
+		{
+			name:                "bind to async-in-progress service instance",
+			asyncForInstances:   true,
+			expectedErrorReason: "ErrorAsyncOperationInProgress",
+		},
+		// TODO: Can't seem to get the test in this state.
+		//{
+		//	name: "unresolved ClusterServiceClass",
+		//},
+		// TODO: Can't seem to get the test in this state.
+		//{
+		//	name: "unresolved ClusterServicePlan",
+		//},
+		{
+			name:                "non-bindable plan",
+			nonbindablePlan:     true,
+			expectedErrorReason: "ErrorNonbindableServiceClass",
+		},
+		//TODO: Can't keep an instance in the not ready state.
+		//{
+		//	name:                "service instance not ready",
+		//	instanceNotReady:    true,
+		//	expectedErrorReason: "ErrorInstanceNotReady",
+		//},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ct := &controllerTest{
+				t: t,
+				skipBindingCreateError: tc.expectedFailure,
+				broker:                 getTestBroker(),
+				instance: func() *v1beta1.ServiceInstance {
+					i := getTestInstance()
+					if tc.instanceClassName != nil {
+						i.Spec.PlanReference.ClusterServiceClassExternalName = *tc.instanceClassName
+					}
+					if tc.instancePlanName != nil {
+						i.Spec.PlanReference.ClusterServicePlanExternalName = *tc.instancePlanName
+					}
+					if tc.nonbindablePlan {
+						i.Spec.PlanReference.ClusterServicePlanExternalName = testNonbindableClusterServicePlanName
+					}
+					if tc.instanceNotReady {
+						newCondition := v1beta1.ServiceInstanceCondition{
+							Type:   v1beta1.ServiceInstanceConditionReady,
+							Status: v1beta1.ConditionFalse,
+						}
+						i.Status.Conditions = []v1beta1.ServiceInstanceCondition{newCondition}
```

Comment:
Any status you add to the ServiceInstance will be overwritten by the API Server. You need to instead set up some condition whereby the Controller will not be able to make the ServiceInstance ready. For example, you could change the response to the Provision request to a non-timeout network error.

Reply:
Looks like that worked! I can add that test back in. Thanks a lot!